src/coreclr/vm/codeman.cpp

```diff
@@ -6273,16 +6360,24 @@ BOOL ReadyToRunJitManager::JitCodeToMethodInfo(RangeSection * pRangeSection,
 
     if (pCodeInfo)
     {
-        pCodeInfo->m_relOffset = (DWORD)
-            (RelativePc - RUNTIME_FUNCTION__BeginAddress(FunctionEntry));
-
         // We are using RUNTIME_FUNCTION as METHODTOKEN
         pCodeInfo->m_methodToken = METHODTOKEN(pRangeSection, dac_cast<TADDR>(FunctionEntry));
 
 #ifdef FEATURE_EH_FUNCLETS
         AMD64_ONLY(_ASSERTE((RawFunctionEntry->UnwindData & RUNTIME_FUNCTION_INDIRECT) == 0));
         pCodeInfo->m_pFunctionEntry = RawFunctionEntry;
 #endif
+        MethodRegionInfo methodRegionInfo;
+        JitTokenToMethodRegionInfo(pCodeInfo->m_methodToken, &methodRegionInfo);
+        if ((methodRegionInfo.coldSize > 0) && (currentInstr >= methodRegionInfo.coldStartAddress))
+        {
+            pCodeInfo->m_relOffset = (DWORD)
```

Comment:
Hmm, can you please explain to me why we don't need <code>RelativePc</code> in this conditional block?

Reply:
`m_relOffset` is meant to be the number of bytes from the beginning of the method as if the function is not split.

So suppose you have a function with 15 bytes, where the last 10 bytes are cold code. So in memory, the method would be layout like this:

```
code byte 0x0 <- RUNTIME_FUNCTION__BeginAddress(FunctionEntry) + ImageBase = methodRegionInfo.hotStartAddress
code byte 0x1
code byte 0x2
code byte 0x3
code byte 0x4

[ ... with some spaces ...]

code byte 0x5 <- `methodRegionInfo.coldStartAddress`
code byte 0x6
code byte 0x7
code byte 0x8
code byte 0x9
code byte 0xa
code byte 0xb
code byte 0xc
code byte 0xd
code byte 0xe
```

To compute `m_relOffset` for hot code, we use this formula.

`RelativePC - RUNTIME_FUNCTION__BeginAddress(FunctionEntry)`

For example, if the current PC is pointing to code byte 0x03, the value is then subtracted by the image base to get RelativePC (i.e. the number of bytes between code byte 0x03 and the image base), and then we subtract the `RUNTIME_FUNCTION__BeginAddress(FunctionEntry)`, which is meant to be the RVA of the function, that will give us 3, which is what we wanted.

To compute the `m_relOffset` for cold code, we do something different.

`methodRegionInfo.coldStartAddress` is an absolute address for where the cold code starts. Assuming the current PC is pointing at code byte 0x09, subtracting the `currentInstr` with `methodRegionInfo.coldStartAddress` gives us 4.

Now if we add the hot code size, which is 5, that gives us 9, which is what we wanted.

For symmetry, since we have `methodRegionInfo.hotStartAddress`, we might be able to write the hot part as:

`pCodeInfo->m_relOffset = (DWORD)(currentInstr - methodRegionInfo.coldStartAddress)`

They should be the same thing, and we will not have the apparent asymmetry in the code.